androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/TwaLauncher.java

```diff
@@ -41,19 +46,41 @@
     private static final int DEFAULT_SESSION_ID = 96375;
 
     public static final FallbackStrategy CCT_FALLBACK_STRATEGY =
-            (context, twaBuilder, providerPackage, completionCallback) -> {
+            (context, twaBuilder, providerPackage, completionCallback, customHeaders) -> {
         // CustomTabsIntent will fall back to launching the Browser if there are no Custom Tabs
         // providers installed.
         CustomTabsIntent intent = twaBuilder.buildCustomTabsIntent();
         if (providerPackage != null) {
             intent.intent.setPackage(providerPackage);
         }
+        applyCustomHeaders(intent.intent, customHeaders);
+
         intent.launchUrl(context, twaBuilder.getUri());
         if (completionCallback != null) {
             completionCallback.run();
         }
     };
 
+    private static void applyCustomHeaders(
+            Intent intent, @Nullable Map<String, String> customHeaders) {
+        if (customHeaders == null) {
+            return;
+        }
+
+        Bundle headerBundle = new Bundle();
+        for (Map.Entry<String, String> header: customHeaders.entrySet()) {
```

Comment:
You can also write this as:

    for (String key: customHeaders.keySet()) {
        String value = customHeaders.get(key);

Up to you though.